zio-http-benchmarks/src/main/scala-2.13/zio/http/benchmarks/EndpointBenchmark.scala

```diff
@@ -84,18 +84,21 @@ class EndpointBenchmark {
 
   private val REPEAT_N = 1000
 
+  import zio.http.codec.PathCodec.int
+
   // # Small Data Request
 
   // API DSL
   val usersPosts =
-    Endpoint
-      .get(PathCodec.literal("users") / PathCodec.int("userId") / "posts" / PathCodec.int("limit"))
+    Endpoint(Method.GET / "users" / int("userId") / "posts" / int("limit"))
       .query(QueryCodec.query("query"))
       .out[ExampleData]
```

Comment:
What's your take on naming something like below?

```scala
Endpoint(Method.GET / "users" / int("userId") / "posts" / int("limit") ^?  "query")
 .body[InputData]
 .response[ExampleData]

```

Reply:
I think the `^` makes it completely undiscoverable

Reply:
I think it's difficult because the body has a response, and the response has a body.

We could be quite verbose: `requestContent`, `responseContent`, or `requestBody`, `responseBody`.

Also these changes prevent us from using `^?`, as now the route pattern needs to be separated from the other codecs to provide unification between this and the lower-level API.

Reply:
I do not like `^?` as a syntax. But I like the idea, to also have query params in the route definition.
I think it is reasonable, to have the option to define it in the routes and get the same benefit of the query param being parsed and the request rejected if a required param is missing.
It seems more user friendly than the alternative of `req.url.queryParams.get("query").flatMap(_.headOption).get` too. And more intuitive too, since the caller puts query params at the same place as a route segment.

Reply:
With this pull request, the query parameters cannot go into the route definition, they must be separated, because endpoints and handlers are unified in `Route`, which requires a `RoutePattern`, which may not mix in query parameters.

So the whole question of `^?` is moot because if this pull request is merged, it will go away forever.

Reply:
Just for my understanding: Why can't query params be an optional part of a route pattern?

Reply:
Because the decision to commit or not to commit occurs at the level of a route pattern: and as far as I know, no web framework allows you to have different handlers based solely on the presence or absence of query parameters.

Reply:
I understand what you mean and guess it is true that no other framework is doing it.
But these frameworks also don't have "reified patterns" ;)

Reply:
There is no theoretical problem doing it, but usually a query parameter is a "setting" for a specific handler, rather than an indication that a whole new handler should handle the same method + path. I think it would be very surprising if dispatch also took into consideration query parameters.

Reply:
Really? If I see the difinition and the query params does not get parsed to an `Option`, is it surprising that when it's missing it can't be handled? And I guess optional handling is need for path segments too or?

Reply:
Maybe it would work, if query params are when defined by default optional, except the user explicitly marks thems as required? 🤔